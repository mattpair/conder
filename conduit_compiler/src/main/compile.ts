
import { Parse} from "./parse";
import { FunctionResolved, TypeResolved, Message, Field } from "./entity/resolved";
import { resolveDeps } from "./resolution/resolveTypes";
import { FileLocation } from "./util/filesystem";
import { Enum, EnumMember } from "./entity/basic";
import { resolveFunctions } from "./resolution/resolveFunction";

export type PartiallyCompiled = Record<string, {proto: string, data: FunctionResolved.File}>
export function compileFiles(files: Record<string, () => string>): PartiallyCompiled {
    const conduits: Parse.File[] = []
    for (const file in files) {
        if (file.endsWith(".cdt")) {
            conduits.push(Parse.extractAllFileEntities(files[file](), new FileLocation(file)))
        }
    }

    const r = resolveDeps(conduits)

    return resolveFunctionsAndProto(r)
} 

function resolveFunctionsAndProto(files: TypeResolved.File[]): PartiallyCompiled {
    const results: PartiallyCompiled = {}
    files.forEach(file => {

        const enums: Enum[] = []
        const messages: Message[] = []
        file.inFileScope.forEach(v => {
            switch(v.kind) {
                case "Enum":
                    enums.push(v)
                    break;
                case "Message":
                    messages.push(v)
                    break;
            }
        })
        let proto = ''
        if (enums.length + messages.length > 0) {
            proto = `
syntax="proto2";
${file.children.Import.map(d => `import "${d.dep.replace(".cdt", ".proto")}";`).join("\n")}
            
${enums.map(printEnum).join("\n\n")}
${messages.map(printMessage).join("\n\n")}
`
        }

        results[`${file.loc.fullname.replace(".cdt", ".proto")}`] = {proto, data: resolveFunctions(file)}
        
    })

    return results       
}

function printEnum(e: Enum): string {
    const mems = printMembers(e.children.EnumMember)
    return `
enum ${e.name} {
${mems}
}
    `
}

function printMembers(m: EnumMember[]): string {
    return m.map((e, index) => `\t${e.name} = ${index + 1};`).join("\n")
}

function printMessage(m: Message): string {
    const fields = printFields(m.children.Field)

    return `
message ${m.name} {
${fields}
}`
}

function printFields(fields: Field[]): string {
    return fields
    .map((f, index) => {
        const p = f.part.FieldType.differentiate()
        return `\t${f.isRequired ? 'required' : 'optional'} ${p.kind === "Primitive" ? p.val : p.name} ${f.name} = ${index + 1};`
    })
    .join("\n")
}