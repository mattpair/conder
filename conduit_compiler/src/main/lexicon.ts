export enum Symbol {
    message="message",
    optional="optional",
    enum="enum",
    EQ="=",
    SEMI=";",
    OPEN_BRACKET="{",
    CLOSE_BRACKET="}",
    COMMA=",",
    double="double",
    float="float",
    int32="int32",
    int64="int64",
    uint32="uint32",
    uint64="uint64",
    sint32="sint32",
    sint64="sint64",
    fixed32="fixed32",
    fixed64="fixed64",
    sfixed32="sfixed32",
    sfixed64="sfixed64",
    bool="bool",
    string="string",
    bytes="bytes",
    NEW_LINE="\\n",
    import="import",
    VARIABLE_NAME="VariableName",
    NUMBER_LITERAL="NumberLiteral",
    STRING_LITERAL="StringLiteral",
    VARIABLE_MEMBER_ACCESS="VarMemberAccess",
    as="as",
    IMPORT_WITH_ALIAS="Import with alias",
}

export const Operators: [
    Symbol.CLOSE_BRACKET, 
    Symbol.OPEN_BRACKET, 
    Symbol.SEMI, 
    Symbol.EQ,
    Symbol.COMMA,
    Symbol.NEW_LINE
] = [
    Symbol.CLOSE_BRACKET, 
    Symbol.OPEN_BRACKET, 
    Symbol.SEMI, 
    Symbol.EQ,
    Symbol.COMMA,
    Symbol.NEW_LINE
]

export type PrimitiveUnion = 
    Symbol.double |
    Symbol.float |
    Symbol.int32 |
    Symbol.int64 |
    Symbol.uint32 |
    Symbol.uint64 |
    Symbol.sint32 |
    Symbol.sint64 |
    Symbol.fixed32 |
    Symbol.fixed64 |
    Symbol.sfixed32 |
    Symbol.sfixed64 |
    Symbol.bool |
    Symbol.string |
    Symbol.bytes

export const Primitives: PrimitiveUnion[] = [
    Symbol.double,
    Symbol.float,
    Symbol.int32,
    Symbol.int64,
    Symbol.uint32,
    Symbol.uint64,
    Symbol.sint32,
    Symbol.sint64,
    Symbol.fixed32,
    Symbol.fixed64,
    Symbol.sfixed32,
    Symbol.sfixed64,
    Symbol.bool,
    Symbol.string,
    Symbol.bytes,
]

export type AnyKeyword = 
Symbol.message |
Symbol.optional | 
Symbol.enum 

export const Keywords: AnyKeyword[] = [
    Symbol.message,
    Symbol.optional,
    Symbol.enum,
]

export enum Dynamic {
    Variable,
    Number
}

const SymbolRegexesMaker: () => Record<Symbol, RegExp> = () => {
    const r = {}
    Operators.forEach(op => {
        r[op] =  new RegExp(`^(?<val>${op})`)
    });

    Primitives.forEach(p => {
        r[p] = new RegExp(`^(?<val>${p})\\s`)
    })

    Keywords.forEach(k => {
        r[k] = new RegExp(`^(?<val>${k})\\s`)
    })
    r[Symbol.NUMBER_LITERAL] = new RegExp(/^(?<val>\d+)/)
    r[Symbol.VARIABLE_NAME] =  /^(?<val>[_A-Za-z]+[\w]*)/
    r[Symbol.STRING_LITERAL] = /^'(?<val>.*)'/
    r[Symbol.VARIABLE_MEMBER_ACCESS] = /^(?<from>[_A-Za-z]+[\w]*)\.(?<type>[_A-Za-z]+[\w]*)/
    r[Symbol.IMPORT_WITH_ALIAS] = /^import +'(?<location>[\w \.\/]*)' +as +(?<alias>[_A-Za-z]+[\w]*)/
    let a = /(?<mycap>[abcdef]*)/
    let b = a.exec("a")
    b.groups
    return r as Record<Symbol, RegExp>
}

export const SymbolToRegex: Record<Symbol, RegExp> = SymbolRegexesMaker()