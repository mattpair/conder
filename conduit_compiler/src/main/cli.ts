import { ConduitBuildConfig, loadBuildConfig } from './config/load';
import * as child_process from 'child_process';
import { generateClients } from './compute/gcp/clients';
import { containerize } from './compute/gcp/deploy';
import {compileFiles} from "./compile"
import { FunctionResolved } from './entity/resolved';
import * as fs from 'fs';
import { deployOnToCluster, destroy, createMedium, MediumState, destroyNamespace } from './deploy/gcp/provisioner';
import { isError } from './error/types';


// This is just a hack for now. I'm the only one running this.
// Revisit once productionizing.
const DEPENDENCY_DIR = '/Users/jerm/ConderSystems/conduit/conduit_compiler/src/main/deps'

// TODO: work across multiple dirs.
function compile(conduits: string[]): Promise<[string[], FunctionResolved.Manifest]>  {
    const toCompile: Record<string, () => string> = {}
    conduits.forEach(c => toCompile[c] = () => fs.readFileSync(`./conduit/${c}`, {encoding: "utf-8"}))
    const [proto, manifest] = compileFiles(toCompile)
    fs.mkdirSync(".proto")
    
    const writes: Promise<string>[] = []
    for (const filename in proto) {
        writes.push(fs.promises.writeFile(`.proto/${filename}`, proto[filename].proto).then(r => filename))
    }
    if (writes.length == 0) {
        console.warn("Did not find any message types in conduit/")
    }

    return Promise.all(writes).then((filenames) => ([filenames,manifest]))
}

const commands: Record<string, () => void> = {
    models() {
        const config = loadBuildConfig()
        if (isError(config)) {
            console.error(config.description)
            return
        }
        let conduits: string[] = []
        try {
            conduits = fs.readdirSync("./conduit/")
        } catch(e) {
            console.error("Unable to find ./conduit/")
            return
        }

        if (conduits.length == 0) {
            console.warn("no files to compile")
            return
        }
        if (!config.dependencies) {
            console.warn("Please specify a dependency in your conduit config so I know where to put models.")
            return
        }
        // TODO: deduplicate
        compile(conduits)
        .then((filenamesAndManifest) => {
            for (const dir in config.dependencies) {
                child_process.execSync(`mkdir -p ${dir}/gen/models`)
                child_process.execSync(`touch ${dir}/gen/models/__init__.py`)
                filenamesAndManifest[0].forEach(p => child_process.execSync(`${DEPENDENCY_DIR}/proto/bin/protoc -I=.proto/ --python_out=${dir}/gen/models ${p} 2>&1`, {encoding: "utf-8"}))    
            }
            
            console.log("done!")
        })
        .catch((e) => console.log("failed.", e))

    },

    async run() {
        const config = loadBuildConfig()
        if (isError(config)) {
            console.error(config.description)
            return
        }
        let conduits: string[] = []
        try {
            conduits = fs.readdirSync("./conduit/")
        } catch(e) {
            console.error("Unable to find ./conduit/")
            return
        }

        if (conduits.length == 0) {
            console.warn("no files to compile")
            return
        }

        const match = /^medium=(?<mediumName>.*)$/.exec(process.argv[3])
        if (match === null || !match.groups.mediumName) {
            console.error(`Need to know which medium to run against`)
            return
        }

        const med: MediumState = JSON.parse(fs.readFileSync(`${STATE_DIR}/mediums/${match.groups.mediumName}.json`, {encoding: "utf-8"}))
        console.log("compiling conduit files")
        compile(conduits)
        .then(async (results) => {
            if (!config.dependencies) {
                const targetDir = '.python'
                child_process.execSync(`mkdir -p ${targetDir}/gen/models`)
                child_process.execSync(`touch ${targetDir}/gen/models/__init__.py`)
                child_process.execSync(`touch ${targetDir}/gen/__init__.py`)
                console.log("generating models from proto")
                results[0].forEach(p => child_process.execSync(`${DEPENDENCY_DIR}/proto/bin/protoc -I=.proto/ --python_out=${targetDir}/gen/models ${p} 2>&1`, {encoding: "utf-8"}))
                console.log("containerizing")
                const image = containerize(results[1], targetDir)
                console.log("deploying to medium")
                const url = await deployOnToCluster(med,results[1].service.name, image, config.project)
    
                if (config.outputClients !== undefined) {
                    console.log("generating clients")
                    config.outputClients.forEach(outputRequest => {
                        child_process.execSync(`mkdir -p ${outputRequest.dir}/gen/models`)
                        child_process.execSync(`touch ${outputRequest.dir}/gen/models/__init__.py`)
                        child_process.execSync(`touch ${outputRequest.dir}/gen/__init__.py`)
            
                        results[0].forEach(p => child_process.execSync(`${DEPENDENCY_DIR}/proto/bin/protoc -I=.proto/ --python_out=${outputRequest.dir}/gen/models ${p} 2>&1`, {encoding: "utf-8"}))
                        generateClients(url, results[1], outputRequest.dir)
                    })
                }
            }
              
            
            console.log("done!")
        })
    },

    async destroy() {
        const match = /^(?<entityType>(medium|deployment))/.exec(process.argv[3])
        
        if (match === null || !match.groups.entityType) {
            console.error("Must specify entity type to delete")
            process.exit(1)
        }
        if (match.groups.entityType === "medium") {
            if (!process.argv[4]) {
                console.error("please specify medium to delete")
                process.exit(1)
            }

            await destroy(JSON.parse(fs.readFileSync(`${STATE_DIR}/mediums/${process.argv[4]}.json`, {encoding: "utf-8"})))
        } else if (match.groups.entityType === "deployment") {
            const config = loadBuildConfig()
            if (isError(config)) {
                console.error(config.description)
                process.exit(1)
            }
            //to-do parameterize
            destroyNamespace(JSON.parse(fs.readFileSync(`${STATE_DIR}/mediums/test-medium.json`, {encoding: "utf-8"})), config.project)
            
        } else {
            console.error(`unable to handle ${match.groups.entityType}`)
        }
    },

    async create() {
        
        if (process.argv[3] === 'medium') {
            if (!process.argv[4]) {
                console.error(`Medium requires a name`)
                return
            }

            await createMedium(process.argv[4]).then((med) => {
                fs.writeFileSync(`${STATE_DIR}/mediums/${process.argv[4]}.json`, JSON.stringify(med))
            })
        } else {
            console.error(`Don't know how to create: ${process.argv[3]}`)
        }
    },

    async has() {
        const match = /^medium=(?<mediumName>.*)$/.exec(process.argv[3])
        if (match === null || !match.groups.mediumName) {
            console.error(`Need to know which medium to look for`)
            process.exit(1)
        }

        if (fs.existsSync(`${STATE_DIR}/mediums/${match.groups.mediumName}.json`)) {
            console.log(`have ${match.groups.mediumName}`)
            process.exit(0)
        }
        console.warn("medium does not exist")
        process.exit(1)
    }
}
const STATE_DIR = child_process.execSync("echo $CONDUIT_STATE_DIR", {encoding: "utf-8"}).trim()
console.log(`State dir: ${STATE_DIR}`)

export function execute() {
    const args = process.argv
    const command = args[2]
    console.log(`Command: ${command}`)
    if (!(command in commands)) {
        console.error(`${command} is invalid.\n\nOptions are ${JSON.stringify(Object.values(commands))}`)
    }

    commands[command]()
}