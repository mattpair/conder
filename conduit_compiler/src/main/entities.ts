import { PrimitiveUnion } from './lexicon';
import { Classified } from './util/classifying';

export type Named = {
    readonly name: string
}

export namespace Resolved {
    
    export enum TypeKind {
        MESSAGE,
        ENUM,
        PRIMITIVE,
    }

    export type MessageOrEnum = Classified<TypeKind.MESSAGE, Unresolved.Message> |
    Classified<TypeKind.ENUM, Enum>

    export type Type = 
    MessageOrEnum |
    Classified<TypeKind.PRIMITIVE, PrimitiveUnion>

    export type Enum = Readonly<{
        members: EnumMember[]
    } & Named>
    
    export type EnumMember = Readonly<Named>
}



export namespace Unresolved {
    export enum FieldKind {
        PRIMITIVE,
        CUSTOM
    }
    
    export type CustomType = {from?: string, type: string}
    export type FieldType = Classified<FieldKind.PRIMITIVE, PrimitiveUnion> | Classified<FieldKind.CUSTOM, CustomType>
    
    
    export type Field = Readonly<{
        isRequired: boolean
        fType: FieldType
    } & Named>
    
    export type Message = Readonly<{
        fields: Field[]
    } & Named>
}
