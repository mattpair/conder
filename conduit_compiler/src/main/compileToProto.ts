
import { tagTokens } from "./parseStep1";
import {parseEntities} from "./parseStep2"
import { Resolved, Unresolved, TypeKind } from "./entities";
import { resolveDeps } from "./resolveDependencies";

export function compileFiles(files: Record<string, () => string>): Record<string, string> {
    const collapsed: Record<string, Unresolved.FileEntities> = {}
    for (const file in files) {
        if (file.endsWith(".cdt")) {
            const entities = parseEntities(tagTokens(files[file]()))
            collapsed[file] = entities 
        }
    }
    const r = resolveDeps(collapsed)
    for (const f in r) {
        r[f].msgs.forEach(m => {
            // console.log(m.name, m.fields.map(field => field.fType.kind !== TypeKind.PRIMITIVE ? field.fType.val() : ""))
        })
    }

    return toProto(collapsed)
} 

function toProto(files: Record<string, Unresolved.FileEntities>): Record<string, string> {
    const results = {}
    for (const key in files) {
        const m: Unresolved.FileEntities = files[key]
        results[key.replace(".cdt", ".proto")] = `
        ${m.enms.map(printEnum).join("\n\n")}
        ${m.msgs.map(printMessage).join("\n\n")}
        `
    }
    return results
}

function printEnum(e: Resolved.Enum): string {
    const mems = printMembers(e.members)
    return `
enum ${e.name} {
${mems}
}
    `
}

function printMembers(m: string[]): string {
    return m.map((e, index) => `\t${e} = ${index + 1};`).join("\n")
}

function printMessage(m: Unresolved.Message): string {
    const fields = printFields(m.fields)

    return `
message ${m.name} {
${fields}
}`
}

function printFields(fields: Unresolved.Field[]): string {
    return fields
    .map((f, index) => `\t${f.isRequired ? 'required' : 'optional'} ${f.fType.kind === TypeKind.PRIMITIVE ? f.fType.val : f.fType.val.type} ${f.name} = ${index + 1};`)
    .join("\n")
}