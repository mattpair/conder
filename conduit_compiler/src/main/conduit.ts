import * as fs from 'fs';
import * as child_process from 'child_process';
import {compileFiles} from "./compileToProto"

// This is just a hack for now. I'm the only one running this.
// Revisit once productionizing.
const DEPENDENCY_DIR = '/Users/jerm/ConderSystems/conduit/conduit_compiler/src/main/deps'

function conduitToProto(conduits: string[]): Promise<string[]>  {
    const toCompile: Record<string, () => string> = {}
    conduits.forEach(c => toCompile[c] = () => fs.readFileSync(`./conduit/${c}`, {encoding: "utf-8"}))
    const protos = compileFiles(toCompile)
    fs.mkdirSync(".proto")
    
    const writes: Promise<string>[] = []
    for (const proto in protos) {
        writes.push(fs.promises.writeFile(`.proto/${proto}`, protos[proto]).then(r => proto))
    }
    if (writes.length == 0) {
        console.warn("Did not find any message types in conduit/")
    }

    return Promise.all(writes)
}

function generateAndDeployHelloWorld() {
    fs.mkdirSync(".deploy/server/server", {recursive: true})
    fs.writeFileSync(".deploy/server/server/__init__.py", "")
    fs.writeFileSync(".deploy/server/server/settings.py",
`
import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '3_+m*a$mw1q4gd22y9j_=ts%%*@_!dfldt2rd05ni#ao0r*e*n'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = False

ALLOWED_HOSTS = ['*']


# Application definition

INSTALLED_APPS = [
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'server.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]


# Database
# https://docs.djangoproject.com/en/3.0/ref/settings/#databases

# DATABASES = {
#     'default': {
#         'ENGINE': 'django.db.backends.sqlite3',
#         'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
#     }
# }


# Password validation
# https://docs.djangoproject.com/en/3.0/ref/settings/#auth-password-validators

# AUTH_PASSWORD_VALIDATORS = [
#     {
#         'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
#     },
# ]


# Internationalization
# https://docs.djangoproject.com/en/3.0/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.0/howto/static-files/

`
    )
    fs.writeFileSync(".deploy/server/server/urls.py",
`
from django.contrib import admin
from django.urls import path
from django.http import HttpResponse


def index(req):
    return HttpResponse("Hello world")

urlpatterns = [
    path('', index),
]
`)
    fs.writeFileSync(".deploy/server/manage.py", 
`
import os
import sys


def main():
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'server.settings')
    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)


if __name__ == '__main__':
    main()
`)

    fs.writeFileSync(".deploy/server/requirements.txt", "Django==3.0")
    fs.writeFileSync(".deploy/Dockerfile",
`
FROM python:3.8-slim-buster

COPY server/ .

RUN pip install -r requirements.txt

EXPOSE  8080

ENTRYPOINT [ "python3", "manage.py", "runserver", "0.0.0.0:8080" ]
`)
    child_process.execSync("docker build -t conder-systems/cloud-run-gen . ", {cwd: ".deploy/"})
    child_process.execSync("docker tag conder-systems/cloud-run-gen us.gcr.io/conder-systems-281115/hello-world-gen", {cwd: ".deploy/"})
}

function main() {
    generateAndDeployHelloWorld()
    // let conduits: string[]
    // try {
    //     conduits = fs.readdirSync("./conduit/")
    // } catch(e) {
    //     console.error("Unable to find ./conduit/")
    //     return
    // }

    // if (conduits.length == 0) {
    //     console.warn("no files to compile")
    // } else {
    //     conduitToProto(conduits)
    //     .then((protos) => {
    //         console.log("done!")
    //         child_process.execSync('mkdir -p python/models')
    //         child_process.execSync('touch python/models/__init__.py')
    //         protos.forEach(p => child_process.execSync(`${DEPENDENCY_DIR}/proto/bin/protoc -I=.proto/ --python_out=python/models ${p} 2>&1`, {encoding: "utf-8"}))
    //     })
    //     .catch((e) => console.log("failed.", e))
    // }    
}

main()

