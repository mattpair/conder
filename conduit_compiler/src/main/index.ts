import * as fs from 'fs';
import {compileFiles} from "./compileToProto"

function conduitToProto(conduits: string[]): Promise<void[]>  {
    const toCompile: Record<string, () => string> = {}
    conduits.forEach(c => toCompile[c] = () => fs.readFileSync(`./conduit/${c}`, {encoding: "utf-8"}))
    const protos = compileFiles(toCompile)
    fs.mkdirSync(".proto")
    
    const writes = []
    for (const proto in protos) {
        console.log(`writing ${proto}`)
        writes.push(fs.promises.writeFile(`.proto/${proto}`, protos[proto]))
    }
    if (writes.length == 0) {
        console.warn("Did not find any message types in conduit/")
    }

    return Promise.all(writes)
}

function main() {
    let conduits: string[]
    try {
        conduits = fs.readdirSync("./conduit/")
    } catch(e) {
        console.error("Unable to find ./conduit/")
        return
    }

    if (conduits.length == 0) {
        console.warn("no files to compile")
    } else {
        conduitToProto(conduits)
        .then(() => console.log("done!"))
        .catch(() => console.log("failed."))
    }    
}

main()

