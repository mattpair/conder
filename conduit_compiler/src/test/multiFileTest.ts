import {compileFiles} from "../main/compileToProto"


function protoCompileTest(description, files: Record<string, string>) {
    test(description, () => {

        const lazyFiles: Record<string, () => string> = {}
        for (const file in files) {
            lazyFiles[file] = () => files[file]
        }
        expect(compileFiles(lazyFiles)).toMatchSnapshot()
    })
} 


protoCompileTest("simple multi file", {
    "conduit_a.cdt": `
    message m1 {
        double d
    }
    `,

    "conduit_b.cdt": `
    import 'conduit_a.cdt' as A

    message m2 {
        double d
    }
    message m1 {
        m2 m
    }
    `
})

protoCompileTest("dependency multi file", {
    "conduit_a.cdt": `
    message m1 {
        double d
    }
    `,

    "conduit_b.cdt": `
    import 'conduit_a.cdt' as A

    message m2 {
        A.m1 m
    }
    `
})