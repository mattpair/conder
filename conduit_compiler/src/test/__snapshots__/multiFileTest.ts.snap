// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circular dependency 1`] = `
"Not all files could be compiled due to circular dependency


        All files: [
  \\"conduit_a.cdt\\",
  \\"conduit_b.cdt\\"
]


        Resolved: []
        "
`;

exports[`dependency multi file 1`] = `
Object {
  "conduit_a.proto": "
        

        
        
message m1 {
	required double d = 1;
}
        ",
  "conduit_b.proto": "
        import \\"conduit_a.proto\\";

        
        
message m2 {
	required m1 m = 1;
}
        ",
}
`;

exports[`dependency subdirs 1`] = `
Object {
  "conduit_a.proto": "
        

        
        
message m1 {
	required double d = 1;
}
        ",
  "conduit_d.proto": "
        import \\"sub/conduit_c.proto\\";

        
        
message D {
	required m2 d = 1;
}
        ",
  "sub/conduit_b.proto": "
        import \\"conduit_a.proto\\";

        
        
message m2 {
	required m1 m = 1;
}
        ",
  "sub/conduit_c.proto": "
        import \\"conduit_a.proto\\";
import \\"sub/conduit_b.proto\\";

        
        
message m2 {
	required m1 m1 = 1;
	required m2 m2 = 2;
}
        ",
}
`;

exports[`simple multi file 1`] = `
Object {
  "conduit_a.proto": "
        

        
        
message m1 {
	required double d = 1;
}
        ",
  "conduit_b.proto": "
        import \\"conduit_a.proto\\";

        
        
message m2 {
	required double d = 1;
}


message m1 {
	required m2 m = 1;
}
        ",
}
`;

exports[`subdir isn't referenced in present dir 1`] = `"Cannot find imported file conduit_c.cdt from file conduit_b.cdt"`;
